pkg/postage/events/events.go

```diff
@@ -0,0 +1,161 @@
+package events
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/bee/pkg/crypto"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/postage"
+)
+
+var (
+	contractAddress      = common.HexToAddress("0x147B8eb97fD247D06C4006D269c90C1908Fb5D54")
+	contractABI          = parseABI("{}")
+	contractEventDigests = digestSig(eventSignatures)
+)
+
+var (
+	eventSignatures = []string{
+		"BatchCreated(bytes32,uint256,address,uint256)",
+		"BatchTopUp(bytes32,uint256)",
+		"BatchDepthIncrease(bytes32,uint256",
+		"PriceUpdate(uint256)",
+	}
+	eventNames = []string{
+		"BatchCreated",
+		"BatchTopUp",
+		"BatchDepthIncrease",
+		"PriceUpdate",
+	}
+)
+
+func eventTypes(i int) postage.Event {
+	switch i {
+	case 0:
+		return &batchCreatedEvent{}
+	case 1:
+		return &batchTopUpEvent{}
+	case 2:
+		return &batchDepthIncreaseEvent{}
+	case 3:
+		return &priceUpdateEvent{}
+	default:
+		return nil
+	}
+}
+
+// event BatchCreated(bytes32 indexed batchId, uint256 initialBalance, address owner, uint256 depth);
+type batchCreatedEvent struct {
+	batchID [32]byte
+	balance *big.Int
+	owner   common.Address
+	depth   *big.Int
+}
+
+func (e *batchCreatedEvent) Update(s postage.EventUpdater) error {
+	return s.Create(e.batchID[:], e.owner[:], e.balance, uint8(e.depth.Uint64()))
+}
+
+// event BatchTopUp(bytes32 indexed batchId, uint256 topupAmount);
+type batchTopUpEvent struct {
+	batchID [32]byte
+	amount  *big.Int
+}
+
+func (e *batchTopUpEvent) Update(s postage.EventUpdater) error {
+	return s.TopUp(e.batchID[:], e.amount)
+}
+
+// event BatchDepthIncrease(bytes32 indexed batchId, uint256 newDepth);
+type batchDepthIncreaseEvent struct {
+	batchID [32]byte
+	depth   *big.Int
+}
+
+func (e *batchDepthIncreaseEvent) Update(s postage.EventUpdater) error {
+	return s.UpdateDepth(e.batchID[:], uint8(e.depth.Uint64()))
+
+}
+
+// event PriceUpdate(uint256 price);
+type priceUpdateEvent struct {
+	price *big.Int
+}
+
+func (e *priceUpdateEvent) Update(s postage.EventUpdater) error {
+	return s.UpdatePrice(e.price)
+}
+
+// parse reifies the event log type as a struct
+//  weakly unsafe in that if there is another event, nil is returned
+func parse(log types.Log) postage.Event {
```

Comment:
at the very least we must check for `Removed` here. Otherwise not only are we not going to catch the reorganisation, we'll apply the removed event twice.